manager/repository/repository.go

```diff
@@ -129,39 +154,48 @@ func (r *mapBasedRepository) CreateDeployment(name string) (*common.Deployment,
 	return d, nil
 }
 
+// AddManifest adds a manifest to the repository and repoints the latest
+// manifest to it for the corresponding deployment.
 func (r *mapBasedRepository) AddManifest(deploymentName string, manifest *common.Manifest) error {
-	err := func() error {
-		r.Lock()
-		defer r.Unlock()
+	r.Lock()
+	defer r.Unlock()
 
-		d, err := r.GetValidDeployment(deploymentName)
-		if err != nil {
-			return err
-		}
+	l, err := r.listManifestsForDeployment(deploymentName)
+	if err != nil {
+		return err
+	}
 
-		l, err := r.listManifestsForDeployment(deploymentName)
-		if err != nil {
-			return err
-		}
+	// Make sure the manifest doesn't already exist, and if not, add the manifest to
+	// map of manifests this deployment has
+	if _, ok := l[manifest.Name]; ok {
+		return fmt.Errorf("Manifest %s already exists in deployment %s", manifest.Name, deploymentName)
+	}
 
-		// Make sure the manifest doesn't already exist, and if not, add the manifest to
-		// map of manifests this deployment has
-		if _, ok := l[manifest.Name]; ok {
-			return fmt.Errorf("Manifest %s already exists in deployment %s", manifest.Name, deploymentName)
-		}
+	d, err := r.GetValidDeployment(deploymentName)
+	if err != nil {
+		return err
+	}
 
-		l[manifest.Name] = manifest
-		d.LatestManifest = manifest.Name
-		r.deployments[deploymentName] = *d
+	l[manifest.Name] = manifest
+	d.LatestManifest = manifest.Name
+	r.deployments[deploymentName] = *d
 
-		return nil
-	}()
+	log.Printf("Added manifest %s to deployment: %s", manifest.Name, deploymentName)
+	return nil
+}
+
+// SetManifest sets an existing manifest in the repository to provided
+// manifest.
+func (r *mapBasedRepository) SetManifest(deploymentName string, manifest *common.Manifest) error {
```

Comment:
As part of the persistent repository work, it might make sense to collapse set/add manifest into a single method. @jackgr